iris.csv를 학습하여 모델을 저장한다.   
svm(svm_iris.h5)   
randomforest(rf_iris.h5)   
knn(knn_iris.h5)   







저장된 모델을 불러와서 예측한다.
- SepalLength: 5.1
- SepalWidth: 3.0
- PetalLength: 1.3
- PetalWidth: 0.2

### iris 데이터 불러오기

In [295]:
import pandas as pd

In [296]:
iris = pd.read_csv("../Data/iris.csv")

iris.head()


SepalLength  SepalWidth  PetalLength  PetalWidth         Name
0          5.1         3.5          1.4         0.2  Iris-setosa
1          4.9         3.0          1.4         0.2  Iris-setosa
2          4.7         3.2          1.3         0.2  Iris-setosa
3          4.6         3.1          1.5         0.2  Iris-setosa
4          5.0         3.6          1.4         0.2  Iris-setosa

### Feature와 Target 분리

In [297]:
iris_data = iris.iloc[:, 0:4] 
iris_target = iris['Name']

print(iris_data.head())
print('--------------------------------')
print(iris_target.head())
print('--------------------------------')
print(iris_data.shape)
print('--------------------------------')
print(iris_target.shape)

   SepalLength  SepalWidth  PetalLength  PetalWidth
0          5.1         3.5          1.4         0.2
1          4.9         3.0          1.4         0.2
2          4.7         3.2          1.3         0.2
3          4.6         3.1          1.5         0.2
4          5.0         3.6          1.4         0.2
--------------------------------
0    Iris-setosa
1    Iris-setosa
2    Iris-setosa
3    Iris-setosa
4    Iris-setosa
Name: Name, dtype: object
--------------------------------
(150, 4)
--------------------------------
(150,)


### train과 test 분리

In [298]:
from sklearn.model_selection import train_test_split

In [299]:
train_input, test_input, train_target, test_target = train_test_split(
    iris_data,
    iris_target,
    random_state=42,
    stratify=iris_target
)

print(train_input.shape)
print('--------------------------------')
print(test_input.shape)
print('--------------------------------')
print(train_target.shape)
print('--------------------------------')
print(test_target.shape)

(112, 4)
--------------------------------
(38, 4)
--------------------------------
(112,)
--------------------------------
(38,)


### SVM

In [300]:
from sklearn.svm import SVC


In [301]:
svm_clf =SVC()

In [302]:
svm_clf.fit(train_input, train_target)
print(svm_clf.score(train_input, train_target))
print(svm_clf.score(test_input, test_target))


0.9732142857142857
0.9210526315789473


In [303]:
# 전체 Data로 모델 만들기
svm_clf.fit(iris_data, iris_target)

SVC()

### AI 저장하기

In [304]:
import joblib

In [305]:
fileName = '../Model/svm_iris.h5'
joblib.dump(svm_clf, fileName)


['../Model/svm_iris.h5']

---

### RandomForest

In [306]:
from sklearn.ensemble import RandomForestClassifier


In [307]:
rf_clf = RandomForestClassifier()


In [308]:
rf_clf.fit(train_input, train_target)
print(rf_clf.score(train_input, train_target))
print(rf_clf.score(test_input, test_target))


1.0
0.9210526315789473


In [309]:
rf_clf.fit(iris_data, iris_target)


RandomForestClassifier()

In [310]:
fileName = '../Model/rf_iris.h5'
joblib.dump(rf_clf, fileName)


['../Model/rf_iris.h5']

---

### kNN

In [311]:
from sklearn.neighbors import KNeighborsClassifier


In [312]:
knn_clf = KNeighborsClassifier()


In [313]:
knn_clf.fit(train_input, train_target)
print(knn_clf.score(train_input, train_target))
print(knn_clf.score(test_input, test_target))


0.9732142857142857
0.9736842105263158


In [314]:
knn_clf.fit(iris_data, iris_target)


KNeighborsClassifier()

In [315]:
fileName = '../Model/knn_iris.h5'
joblib.dump(knn_clf, fileName)



['../Model/knn_iris.h5']

SVM:   
0.9732142857142857  
0.9210526315789473   

RF:  
1.0  
0.8947368421052632  
 
kNN:  
0.9732142857142857  
0.9736842105263158  

> kNN이 적절 (5프로 이내가 좋음)

---

### 저장된 모델 불러와 예측
- SepalLength: 5.1
- SepalWidth: 3.0
- PetalLength: 1.3
- PetalWidth: 0.2

In [316]:
# 데이터 만들기
pred_data = {
    'SepalLength': [5.1],
    'SepalWidth': [3.0],    
    'PetalLength': [1.3],
    'PetalWidth': [0.2]
}

df = pd.DataFrame(pred_data)
df

SepalLength  SepalWidth  PetalLength  PetalWidth
0          5.1         3.0          1.3         0.2

In [317]:
# svm
svm_clf = joblib.load('../Model/svm_iris.h5')
print(svm_clf.predict(df))

# rf
rf_clf = joblib.load('../Model/rf_iris.h5')
print(rf_clf.predict(df))

# knn
knn_clf = joblib.load('../Model/knn_iris.h5')
print(knn_clf.predict(df))


['Iris-setosa']
['Iris-setosa']
['Iris-setosa']
